lib/noid/template.rb

```diff
@@ -15,93 +16,105 @@ def mint(n)
       str
     end
 
+    ##
+    # Is the string valid against this template string and checksum?
+    # @param [String] str
+    # @return bool
     def valid?(str)
-      return false unless str[0..prefix.length] == prefix
-
-      if generator == 'z'
-        str[prefix.length..-1].length > 2
-      else
-        str[prefix.length..-1].length == characters.length
-      end
-
-      characters.split('').each_with_index do |c, i|
-        case c
-        when 'e'
-          return false unless Noid::XDIGIT.include? str[prefix.length + i]
-        when 'd'
-          return false unless str[prefix.length + i] =~ /\d/
-        end
-      end
-
-      return false unless checkdigit(str[0..-2]) == str.split('').last if checkdigit?
-
+      match = validation_regex.match(str)
+      return false if match.nil?
+      return checkdigit(match[1]) == match[3] if checkdigit?
       true
     end
 
     ##
-    # identifier prefix string
-    def prefix
-      @prefix ||= @template.split('.').first
+    # calculate a checkdigit for the str
+    # @param [String] str
+    # @return [String] checkdigit
+    def checkdigit(str)
+      Noid::XDIGIT[str.split('').map { |x| Noid::XDIGIT.index(x).to_i }.each_with_index.map { |n, idx| n * (idx + 1) }.inject { |sum, n| sum + n } % Noid::XDIGIT.length]
     end
 
     ##
-    # identifier mask string
-    def mask
-      @mask ||= @template.split('.').last
+    # minimum sequence value
+    def min
+      @min ||= 0
     end
 
     ##
-    # generator type to use: r, s, z
-    def generator
-      @generator ||= mask[0..0]
+    # maximum sequence value for the template
+    def max
+      @max ||= if generator == 'z'
+                 nil
+               else
+                 size_list.inject(1) { |total, x| total * x }
+               end
     end
 
+    protected
+
     ##
-    # sequence pattern: e (extended), d (digit)
-    def characters
-      @characters ||= begin
-                        if checkdigit?
-                          mask[1..-2]
-                        else
-                          mask[1..-1]
-                        end
-                      end
+    # A noid has the structure (prefix)(code)(checkdigit)
+    # the regexp has the following captures
+    #  1 - the prefix and the code
+    #  2 - the changing id characters (not the prefix and not the checkdigit)
+    #  3 - the checkdigit, if there is one. This field is missing if there is no checkdigit
+    def validation_regex
+      @validation_regex ||= begin
```

Comment:
I whipped up two alternative regex builders, one based on an array of regex that are mapped together (per your example) and the other a multi-line regex.  The former wouldn't work cleanly because some segments of the builder generate syntactically invalid regex. In particular, changing line 67 to `pattern_list << /))/` causes Ruby to barf. The latter was way uglier than the string builder. I'm back to the string builder but willing to entertain alternative suggestions if you really don't want this as is.